### PR TITLE
fix(adapters): give psmux a per-window option channel

### LINKED ISSUE #310: psmux: per-window user options are session-scoped - the ctl-window project guard cannot discriminate and the parked-return mechanism is inert
## Summary

psmux has exactly **one user-option scope per server**. Every `set-option -w` write lands in that one map and every `-w` read of an `@`-prefixed name returns `''`, regardless of target. Two bmad-loop mechanisms ride per-window user options and are therefore broken on psmux **today**, independently of any window-id work:

- **`PROJECT_OPTION` cross-project prune scoping does not discriminate.** `_ctl_window_candidates` reads `runs.PROJECT_OPTION` as a `list-windows` field to leave another project's ctl window alone (`tui/launch.py:214`, tag written at `:310`). On psmux the stored value is session-scoped and `#{@bmad_project}` expands to the same value for *every* row, so the guard cannot tell one project's window from another's on the shared `bmad-loop-ctl` session.
- **`RETURN_OPTION` — the parked-window return mechanism — is fully inert.** The trailer reads it with `show-options -wqv` (`adapters/psmux_backend.py:107`, base at `adapters/tmux_base.py:231`) and `return_attached_client` does the same at `tui/launch.py:146`. That read always yields `''` on psmux, so neither the `detach` nor the `switch-client` branch ever fires: an attached client is never returned to its origin.

This was split out of #291 by scope review — #291 is about *id routing*, this is about needing a different *state channel*. Filing separately rather than growing that diff.

## Verified in the psmux source at tag `v3.3.7`

`v3.3.7` is the tag matching the installed build (`05cc5d4`). Read via `git show v3.3.7:<path>`, since a working checkout a few commits past the tag touches some of these files.

- `src/server/options.rs:4` — `is_window_option` is a hard `matches!` allowlist of exactly **14 builtin names** (`automatic-rename`, `monitor-activity`, `remain-on-exit`, the `window-status-*` family, `main-pane-*`, `window-size`). No `@`-prefixed name can ever be in it.
- `src/server/options.rs:155` — `get_window_option_value_for` returns `String::new()` **immediately** when `!is_window_option(name)`, before ever consulting the user-options map. This is the whole read path for `show-options -wqv`, so a `@`-prefixed `-w` read is empty in every case: targeted or untargeted, in-pane or out.
- `src/types.rs` — the `Window` struct has **no options storage of any kind** (`root`, `active_path`, `name`, `id`, flags, layout, panes — no options map).
- `src/types.rs:676` — `user_options: HashMap<String, String>` lives on `AppState`, i.e. one map per server, and psmux runs one server per session.
- `src/server/options.rs:529,538` — every `set-option` write path inserts into `app.user_options` with no scope or target recorded.
- `src/server/options.rs:123` — the *session-scope* read (`show-options -qv`, no `-w`) **does** consult `app.user_options`. So a session-scoped user option round-trips correctly; only the `-w` surface is dead.

Live-probed on stock 3.3.7 alongside the source read: after `set-option -w -t A:@2 @opt vid`, every targeted `-wqv` read returns `''` while `show-options -qv -t A @opt` returns `vid`, and `list-windows -t A -F '#{window_name}=[#{@opt}]'` prints that one value for every window of the session.

## This is a model boundary, not a regression to wait out

Same family as the per-server window ids in #254 — a deliberate property of psmux's one-server-per-session model, not a defect an upstream release will fix under us. The maintainer states it directly in psmux/psmux#321: options "only exist at the global level in psmux" (the `-w` read surface exists for libtmux/tmuxp compatibility, backed by the builtin allowlist plus a global fallback for a few names).

The only genuinely bug-shaped edges are cosmetic by comparison — `set-option -w -t <target>` is silently accepted with rc 0 while the target is ignored, and a write is unreadable through the same flags that set it — and fixing those would make the calls *error*, not make per-window options exist. So the fix belongs on our side.

## Candidate channels

Both work on stock 3.3.7 (round-trip live-verified):

1. **Session-scoped user option with the window id encoded in the option _name_** — e.g. `@bmad_return_%40N` / `@bmad_project_%40N`, read in-pane after a `display-message -p '#{window_id}'`. Stays inside the multiplexer; needs a cleanup story so keys do not accumulate on the long-lived shared ctl session.
2. **State keyed by window id outside the multiplexer entirely** — a run-dir sidecar, as the engine already does for its own state. No psmux limits, but it puts ctl-window state somewhere the parked trailer (a shell one-liner) has to read from.

The backend override seam established by #290 is the right place to hide whichever channel is chosen, so the tmux path keeps using real window options.

## Ordering

Do this **after** #291. Channel 1 keys on the window id read from the seam #291 qualifies; landing it first would bind the option names to the bare id form and force a migration.

Refs: #291, #290, #254, psmux/psmux#321

### PR DESCRIPTION
**Pull request by dracic**
Closes #310

## What

psmux keeps every user option in one per-server map and short-circuits every `-w` read of an `@`-prefixed name to `""` before that map is consulted. Both mechanisms riding per-window options were therefore broken: the ctl-window `PROJECT_OPTION` prune guard read the same bled value for every row (a prune in one project could `kill-window` another project's window), and the `RETURN_OPTION` parked-return mechanism was fully inert.

- The three window-option verbs are overridden onto psmux's one working channel: a **session-scoped** user option keyed by the full window id (`@bmad_project_@3` for window `@3`), every out-of-pane write and read routed with an explicit `-t <session>`. Builtin (non-`@`) options fall through to the base `-w` path. No call-site changes.
- `list_windows` drops `@`-prefixed fields from the psmux format string (the `#{@name}` bleed) and fills those columns per row from one batched `show-options -q` listing. A failed read poisons the column instead of reading as "untagged" — `""` is the kill-eligible direction.
- Values that cannot survive psmux's lossy CLI→server hop verbatim are refused loudly (`_transportable`). That includes standalone `;`/`\;` tokens: the one-shot CLI path splits them without quote-awareness and hands the remainder to the server as a command — filed upstream as psmux/psmux#499 with an execution proof.
- The parked trailer self-resolves its own window id in-pane (`-t $env:TMUX_PANE`) and reads/unsets its id-suffixed key. The switch-client leg composes the correct verb but stays inert until psmux/psmux#483 lands upstream; the detach leg works today.
- Key lifecycle: `kill_window` sweeps that window's `_@N` session keys before the kill, and a launch-time orphan reconcile covers the Enter-dismiss exit that `kill_window` never sees (snapshot-before-listing ordering so a freshly minted window can never read as an orphan).
- tmux untouched: its `-w` argv is byte-identical to before (divergence pinned by test).

Every psmux behavior relied on here was read from the psmux source at tag `v3.3.7` (the installed build) and the transport edge cases were smoke-tested live against a scratch session on stock 3.3.7.

## Verification

- `pytest tests/test_psmux_backend.py tests/test_multiplexer.py -q` → 134 passed (incl. ablation-checked degrade/refusal gates and the tmux divergence pins)
- full suite: no new failures vs. the local baseline
- `trunk check` clean, `pyright@1.1.411` clean on the touched modules

## Summary by CodeRabbit

* **Bug Fixes**
  * Fixed psmux window targeting and per-window option handling.
  * Restored returning detached clients to their originating windows.
  * Prevented option values from being corrupted during transport; unsupported values are now rejected.
  * Added cleanup for stale window options and improved handling when windows are removed or parked.

* **Documentation**
  * Clarified psmux’s session-scoped window model, option behavior, and transport limitations.

### CODE REVIEW

File: src/bmad_loop/adapters/psmux_backend.py [hunk 1]
```diff
@@ -232,23 +266,353 @@ def new_parked_window(
...
+        if not option.startswith("@"):
+            super().set_window_option(target, option, value)
+            return
+        if not self._transportable(value):
+            print(
+                f"warning: set-option {option} skipped — value does not survive "
+                "psmux's control-line transport verbatim; the option stays unset",
+                file=sys.stderr,
+            )
+            return
```
[the remark this answers is not in the document]
Comment (dracic) [reply src/bmad_loop/adapters/psmux_backend.py:504]: Fixed in 2550377. The refusal branch now frees the key via `unset_window_option` (which itself warns if the unset cannot land), and the warning is reworded to "the key is freed and the option reads as unset". Both refusal tests now pin that the only verb a refused write emits is the single `set-option -u` — so a stale parked return target can no longer survive a refused rewrite.

File: src/bmad_loop/adapters/psmux_backend.py [hunk 2]
```diff
@@ -232,23 +266,353 @@ def new_parked_window(
...
+        # own reads strip/Trim, so the value could never read back equal.
+        if not value or value.startswith("-") or any(c.isspace() and c != " " for c in value):
+            return False
+        if value != value.strip():
+            return False
+        if " " in value:  # will be double-quoted by the psmux client
+            if '"' in value or "\\\\" in value or value.endswith("\\"):
+                return False
+            return all(tok not in (";", "\\;") for tok in value.split())
+        return not any(c in value for c in ";'\"")
```
[the remark this answers is not in the document]
Comment (dracic) [reply src/bmad_loop/adapters/psmux_backend.py:464]: Checked against the psmux source at tag `v3.3.7` rather than assumed: the tokenizer's backslash-collapse branch is `else if c == '\' && in_double_quotes` (commands.rs:690) — outside double quotes a backslash falls through to the literal-push arm, with an in-source comment noting backslash is kept literal because psmux is Windows-native and it is the normal path separator. The client only wraps a value in double quotes when it contains an ASCII space (main.rs:2955), so a spaceless `\srv\share` is sent unquoted and stored verbatim — no collapse, and the prune's read-back equality holds.

To make that durable instead of a review-thread fact, 2550377 adds `\srv\share` to `test_accepted_values_round_trip_through_the_listing_parse` (asserts both `_transportable` acceptance and byte-identical read-back through the listing parse) and extends the `_transportable` comment to state why the unspaced branch deliberately does not ban `\`.

**Comment by dracic** [comment]
Heads-up for whoever merges this, from the #227 / PR #316 review — a constraint this PR *inherits* rather than causes.

While the option channel is inert, `tui.launch.return_attached_client` reads an empty `RETURN_OPTION` on psmux and answers `False` before it ever reaches a `switch-client` or `detach-client` call. **This PR removes that shield.** Once the channel works, the parked-return path reaches those verbs — and on psmux both report execution, not effect:

- at `v3.3.7` the whole `switch-client` arm is `send_control(cmd)?; return Ok(())` (no reply for any flag, `-t` included), and `detach-client` exits nonzero only when the session server is unreachable, so it succeeds with zero clients attached;
- current `main` adds a response path for `switch-client -t` alone, leaving `-l` and `detach-client` exit-0-regardless.

PR #316 just made `return_attached_client` trust those booleans to decide whether to clear the return option. A vacuous `True` therefore clears it and reports a hand-back that never happened — #227's exact symptom, on psmux, reachable only after this lands.

Filed as **#317** with the source evidence and three candidate directions, so nothing here needs to change: this is a note that the option channel's first live consumer arrives with a known-untrustworthy signal underneath it. Worth adding to the #313 safety-core checklist rather than treating as a blocker for this merge.

**Comment by pbean** [comment]
Reviewed against #310, the design doctrine in AGENTS.md/CONTRIBUTING, and the POSIX path. **The approach is the one to take and the scope is right** — I pushed four small corrections onto the branch rather than sending it back, and rebased onto `main`.

### What I checked, and what held

- **Channel choice.** #310 offered two; picking channel 1 and rejecting the run-dir sidecar because `set_window_option` / `new_parked_window(…, return_opt)` carry no run-dir — so a sidecar forces an ABC change that also touches tmux — is the right call, and it keeps psmux divergences behind the #290/#291 override seam as designed.
- **No core edits.** The only non-psmux source change was two docstrings in `adapters/multiplexer.py` (one of which I removed, below).
- **Linux/macOS untouched.** `PsmuxMultiplexer` registers for `win32` only; every override is in the leaf; the base's `-w` argv is byte-identical and now pinned by two tmux tests. Builtin (non-`@`) options still fall through to `super()`, so the 14 real window options keep working.
- **Invariants.** No policy field, env var, `--json` contract or `ExitCode` touched. Argv stays in a sanctioned mux-backend module. Best-effort verbs warn-never-raise; the launch-time sweep swallows `SubprocessError | OSError | TmuxError` so a mint can never fail on reconcile.
- **Tests aren't vacuous.** I ablated each refusal/degrade gate — the `degraded` branch, the three `scope is None` guards, `if not live: return`, and the `_transportable` call — and every one produced a failure. The order-pin test inverts correctly too: swapping the two listings in `_sweep_orphan_keys` fails it.

### What I changed (3 commits)

**1. `fix(adapters): degrade psmux option columns to unset (#310 review)`** — the one substantive finding.

`_OPT_READ_FAILED` guarded a hazard the caller's untagged fallback already covers. `_ctl_window_candidates` claims an untagged window only when its run dir exists **under this project**, and a run id is `%Y%m%d-%H%M%S-<4 hex>` — so `""` is safe in both directions: ours → correctly pruned, theirs → no run dir here → skipped anyway. The poison's only real effect was to skip pruning *our own* dead window on a transient failure.

It was also inconsistent with the channel's own documented hole: `_scoped_options`' docstring records that a psmux handler can answer the listing **empty-with-success while keys exist**, and `list_windows` filled `""` on that path regardless — so the exact hazard the sentinel existed to prevent was reachable, silently, through a path the PR documents. Both halves now degrade the same way, and the cross-backend rule ("an `@` column may carry a backend-private opaque marker") comes back out of the ABC. I kept the `set_window_option` docstring paragraph — "the contract is the keying, not the storage" is the seam rule this fix genuinely relies on.

I also anchored the surviving warning the way #291 taught us: it is visible on the CLI prune (including the `--dry-run` your comment cites) but **not** under the TUI, which captures `sys.stderr` for the app's whole run — same deliberate ceiling as the `select_window` note.

**2. `fix(adapters): resolve psmux option scope before the transport gate`** — a refused value re-entered `unset_window_option` to free the stale key, so an unroutable target warned twice: once about the transport, once about a `set-option -u` the caller never issued. Scope resolves first (an unroutable target has nothing to write *and* nothing to free), then the value is gated, and the key is freed through the already-resolved scope. The two mutating verbs now share one spawn-and-warn body.

**3. `docs(adapters): pin the spaceless-apostrophe transport refusal`** — the docs cited only "UNC and spaced", but since the client quotes only on an ASCII space and `'` is literal inside those quotes, `C:\Users\O'Brien\dev` is **refused** while `C:\Users\O'Brien Files\dev` is accepted. The inversion is correct per your transport model, but an apostrophe in a Windows home is far more common than a UNC+spaced project path, so it is now named in the docs and pinned by a parametrize case. Also pointed the module docstring at the channel block instead of restating the model, and replaced a stray word in the `kill_window` cost note.

### Left for you

- **Live verification is the gap CI cannot close.** Every test fakes `tmux_base.subprocess.run`, so green proves argv composition only. Three facts the fix rests on are only as good as your live probes — most load-bearing being that **psmux sets `TMUX_PANE` in every pane**: if it doesn't, `$wid` is empty and the whole return mechanism stays inert, which is half of #310. `pwsh` isn't on my host, so I could not even parse the composed trailer. If you can paste the before/after of a real launch→attach→exit (detach leg returning the client) and a two-project `bmad-loop cleanup --dry-run`, that closes it.
- **Size.** 897 changed lines against CONTRIBUTING's 800 maximum. The source is 388 — the tests are what push it over, and they're good tests, so I'm not asking you to cut them; flagging it so the call is explicit rather than silent.
- **Follow-up filed: #320.** `session_options()` reads `#{@bmad_project}` through `list-sessions -F` — the same single-map expansion this PR calls the bug on the window side — and `psmux_backend` overrides neither it nor `set_session_option`, so the **agent**-session prune may be mis-scoped the same way (and the write is ungated by `_transportable`). Out of #310's scope, deliberately not grown into this diff; the issue lists the three things to probe before any code.

Local: full suite green apart from two pre-existing `test_module_skills_sync` failures that also fail on `main` (gitignored seed drift, unrelated to this branch); `trunk check` clean; `pyright@1.1.411` clean.

**Comment by pbean** [comment]
## Live verification — psmux 3.3.7 (05cc5d4), native Windows

Ran the whole thing on a Windows VM against `99385fe`, then repeated the prune half against `d3c6f5e` for the before/after. Zero LLM tokens; every check is raw psmux or the backend driven directly. **All five areas pass**, and two assumptions the code rests on are now probed rather than assumed.

### The `TMUX_PANE` claim holds — this was the load-bearing one

```
TMUX=[/tmp/psmux-7556/default,56590,0]
TMUX_PANE=[%1]
psmux display-message -p -t $env:TMUX_PANE '#{window_id}'  ->  @1
psmux display-message -p '#{window_id}'                    ->  @1
```

Also confirmed separately: the trailer's **untargeted** in-pane read reaches the session's map, so `$TMUX` routing is real and the missing `-t <session>` in `_parked_trailer` is correct. Replayed by hand in a pane against a key set from outside — `ret=[detach]`, matching the targeted read.

### The per-window channel discriminates

```
ids: psmux-probe:@2 psmux-probe:@3
raw listing: {'@bmad_project_@3': 'C:\\proj\\beta', '@bmad_project_@2': 'C:\\proj\\alpha'}
row: ('psmux-probe:@1', 'pwsh',  '')
row: ('psmux-probe:@2', 'alpha', 'C:\\proj\\alpha')
row: ('psmux-probe:@3', 'beta',  'C:\\proj\\beta')
keys after kill_window(@2): {'@bmad_project_@3': 'C:\\proj\\beta'}
```

Rows carry distinct values, and `kill_window` freed that window's key and only that one.

### `_transportable` is exactly calibrated — no corruption, no false refusals

Round-tripped ten values through a real write + read. Every one is either byte-identical or refused; **nothing stored corrupted**:

| value | result |
|---|---|
| `C:\proj\normal` | OK |
| `C:\Users\Some User\p` | OK |
| `C:\Users\O'Brien Files\proj` | OK |
| `a; b` | OK |
| `two  spaces` | OK |
| `\\srv\share` | OK |
| `C:\Users\O'Brien\dev` | refused |
| `a ; b` | refused |
| `\\srv\share My Proj` | refused |
| `C:\dir with space\` | refused |

The accept/refuse split matches your model exactly, including the counter-intuitive pair — spaced `O'Brien Files` accepted, spaceless `O'Brien\dev` refused.

### The parked return actually returns now

Real `new_parked_window`, attach, `select-window` to it, Enter at the park prompt:

```
window: parkprobe:@2
psmux show-options -qv -t parkprobe "@bmad_return_pane_@2"  ->  detach
psmux attach -t parkprobe        # ... Enter at the banner -> dropped to the host prompt
psmux has-session -t parkprobe; echo "rc=$LASTEXITCODE"     ->  rc=0
psmux show-options -q -t parkprobe | Select-String bmad     ->  (nothing)
```

Client detached, session survived, key freed. The `has-session` line matters: a read against a dead session doesn't fail on psmux, it falls through the most-recent-session routing and answers from another server — so an option read is only evidence while the session is provably alive.

The `switch-client` leg is untested and stays inert per psmux/psmux#483, as the PR says.

### Cross-project prune — and the "before" is worse than #310 documented

On `99385fe`, two projects sharing `bmad-loop-ctl`:

```
@bmad_project_@2 "C:\tmp\projA"
@bmad_project_@3 "C:\tmp\projB"

cleanup --dry-run --project C:\tmp\projA  ->  would close ctl window run-20260101-000000-aaaa
cleanup --dry-run --project C:\tmp\projB  ->  would close ctl window run-20260102-000000-bbbb
```

Same scenario on `d3c6f5e` (has #291, not #310) — the map collapses to one key, last writer wins:

```
@bmad_project "C:\tmp\projB"

cleanup --dry-run --project C:\tmp\projA  ->  nothing to clean up
cleanup --dry-run --project C:\tmp\projB  ->  would close ctl window run-20260101-000000-aaaa
                                              would close ctl window run-20260102-000000-bbbb
```

Both failure directions, live: a project **cannot prune its own window**, and the other project **would kill it**. The issue only documented the second.

The pre-fix return is reproducible too — on `d3c6f5e`, Enter at a parked window leaves the client attached and drops it to window 0, because the `-wqv` read yields `''` and neither branch fires.

### Two assumptions, now probed rather than assumed

**Window ids are not recycled** — the sweep's correctness comment holds. If they were, a new window would inherit a dead one's stale key:

```
@1  ->  new-window a  ->  @1 @2  ->  kill @2  ->  new-window b  ->  @1 @3
```

**`new-session` on an existing session errors** (`duplicate session: dup`, rc 1), so `new_session`'s post-create `has_session` belt is meaningful rather than papering over a silent no-op.

### Two cosmetic notes, neither blocking and neither yours

- `psmux list-windows` renders `run-20260101-000000-aaaa-` with a trailing `-` — that is psmux's default format appending window flags, not a corrupted name. bmad-loop passes its own `-F #{window_name}`, and `cleanup` printed the clean name, so nothing is affected.
- `bmad-loop mux` renders a broken table on a psmux host: `psmux -V` prints two lines and `cli.py:407` drops `r.version` straight into a `ljust` column. The `tmux` row also reports psmux's version, since psmux ships a `tmux`-named binary — harmless for selection today because tmux's platform gate excludes win32. Pre-existing and unrelated to this PR; I'll file it separately.

That closes the verification gap I flagged. From my side this is ready to merge.
